lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp

```diff
@@ -892,6 +896,128 @@ class SubtractOpConversionPattern
   }
 };
 
+class AllReduceOpConversionPattern
```

Comment:
I am currently working on adding a workaround pass in TTNN here:
https://github.com/tenstorrent/tt-mlir/pull/1402

This seems to be one of the workarounds that should end in the workaround framework. I think we want to create a TTNN all-reduce op, and decompose it as a part of a workaround pass in other TTNN ops and do that in a proper and single place where all workarounds should exist.

Reply:
Potentially yes, Wooseok and I had this discussion early last week.  This one is a bit of a gray area because the TTNN op AllReduce doesn't exist and it's not clear when it will exist.  I think they have plans to eventually add it, but I guess this boils down to the TTNN dialect and if we want it to accurately reflect the API that currently exists or if we're OK introducing ops that don't exist with the expectation that they will be decomposed.

Wooseok's approach seems reasonable because it maintains the loose invariance that all TTNN dialect ops are backed by real TTNN API.

Reply:
@sdjordjevicTT are you OK to land this as is because of the reason I outlined above or do you think should definitely be part of a workaround?

Reply:
If we don't expect AllReduce to be included in TTNN, then that's acceptable to me. However, rather than introducing a complex decomposition during a conversion pass, could we move this decomposition to a TTIR decomposition pass instead? We agreed that we want to keep TTNN conversions as simple as possible, and the move to decomposition pass will allow us to achieve that. In this case, I rather see that as a right solution rather than a workaround.

Reply:
@sdjordjevicTT  I thought about splitting the op at the TTIR level. Then I realized that it is the problem of TTNN, so I believe that we should put this inside of TTNN boundary such that we can reduce the impact of code change if ttnn improves their API in the future. That was my thoughts.

Reply:
Yeah I agree with @wooseokTT, this is a TTNN limitation, splitting up AllReduce in TTIR isn't a strict simplification and is leaking a TTNN detail into TTIR, I'd rather introduce the dummy TTNN AllReduce.  Eventually I think TTNN will have an AllReduce I just have no idea when it'll happen esp since this is a functional solution.

Reply:
My preference in order between the approaches:
- Merge PR change as is
- Create ttnn.all_reduce, use WA pass to lower it to ttnn.scatter/gather
- Do the decomp in TTIR